### PR TITLE
Add Feature-Policy header

### PR DESCRIPTION
**Pull request by rullzer**

### CODE REVIEW

**Comment by rullzer** [comment]
Review time :)

This is not all the features. But other things will be added later in steps.

File: lib/public/AppFramework/Http/EmptyFeaturePolicy.php [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	protected $paymentDomains = null;
+
+	/**
+	 * Allows to use autoplay from a specific domain. Use * to allow from all domains.
+	 *
+	 * @param string $domain Domain to whitelist. Any passed value needs to be properly sanitized.
+	 * @return $this
+	 * @since 17.0.0
+	 */
+	public function addAllowedAutoplayDomain(string $domain) {
```
Comment (georgehrke) [inline lib/public/AppFramework/Http/EmptyFeaturePolicy.php]: `:self`

File: lib/public/AppFramework/Http/EmptyFeaturePolicy.php [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	}
+
+	/**
+	 * Allows to use the camera on a specific domain. Use * to allow from all domains
+	 *
+	 * @param string $domain Domain to whitelist. Any passed value needs to be properly sanitized.
+	 * @return $this
+	 * @since 17.0.0
+	 */
+	public function addAllowedCameraDomain(string $domain) {
```
Comment (georgehrke) [inline lib/public/AppFramework/Http/EmptyFeaturePolicy.php]: `:self`

File: lib/public/AppFramework/Http/EmptyFeaturePolicy.php [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	}
+
+	/**
+	 * Allows the full screen functionality to be used on a specific domain. Use * to allow from all domains
+	 *
+	 * @param string $domain Domain to whitelist. Any passed value needs to be properly sanitized.
+	 * @return $this
+	 * @since 17.0.0
+	 */
+	public function addAllowedFullScreenDomain(string $domain) {
```
Comment (georgehrke) [inline lib/public/AppFramework/Http/EmptyFeaturePolicy.php]: `:self`

File: lib/public/AppFramework/Http/EmptyFeaturePolicy.php [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+	}
+
+	/**
+	 * Allows to use the geolocation on a specific domain. Use * to allow from all domains
+	 *
+	 * @param string $domain Domain to whitelist. Any passed value needs to be properly sanitized.
+	 * @return $this
+	 * @since 17.0.0
+	 */
+	public function addAllowedGeoLocationDomain(string $domain) {
```
Comment (georgehrke) [inline lib/public/AppFramework/Http/EmptyFeaturePolicy.php]: `:self`

File: lib/public/AppFramework/Http/EmptyFeaturePolicy.php [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+	}
+
+	/**
+	 * Allows to use the microphone on a specific domain. Use * to allow from all domains
+	 *
+	 * @param string $domain Domain to whitelist. Any passed value needs to be properly sanitized.
+	 * @return $this
+	 * @since 17.0.0
+	 */
+	public function addAllowedMicrophoneDomain(string $domain) {
```
Comment (georgehrke) [inline lib/public/AppFramework/Http/EmptyFeaturePolicy.php]: `:self`

File: lib/public/AppFramework/Http/EmptyFeaturePolicy.php [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+	}
+
+	/**
+	 * Allows to use the payment API on a specific domain. Use * to allow from all domains
+	 *
+	 * @param string $domain Domain to whitelist. Any passed value needs to be properly sanitized.
+	 * @return $this
+	 * @since 17.0.0
+	 */
+	public function addAllowedPaymentDomains(string $domain) {
```
Comment (georgehrke) [inline lib/public/AppFramework/Http/EmptyFeaturePolicy.php]: `:self`

**Review by georgehrke — DISMISSED** [review]
Code looks good, just some minor changes.

To test just make sure that the header is there?

**Comment by rullzer** [comment]
Yes check the header is there.
You could check with talk. That should fail hard now until I provide a patch there ;)

File: lib/public/AppFramework/Http/EmptyFeaturePolicy.php (hunk 1 above)
Comment (rullzer) [reply lib/public/AppFramework/Http/EmptyFeaturePolicy.php]: fair enough

File: lib/private/AppFramework/DependencyInjection/DIContainer.php [hunk 7]
```diff
@@ -231,6 +231,11 @@ public function __construct($appName, $urlParams = array(), ServerContainer $ser
 					$server->query(OC\Security\CSRF\CsrfTokenManager::class)
 				)
 			);
+			$dispatcher->registerMiddleware(
+				new OC\AppFramework\Middleware\Security\FeaturePolicyMiddleware(
+					$server->query(OC\Security\FeaturePolicy\FeaturePolicyManager::class)
```
Comment (ChristophWurst) [inline lib/private/AppFramework/DependencyInjection/DIContainer.php]: I guess you're just following the style of the other code but could you just query the middleware instance?

File: lib/private/Security/FeaturePolicy/FeaturePolicyManager.php [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	 *
+	 */
+	public function mergePolicies(FeaturePolicy $defaultPolicy,
+								  EmptyFeaturePolicy $originalPolicy): FeaturePolicy {
+		foreach ((object)(array)$originalPolicy as $name => $value) {
+			$setter = 'set' . ucfirst($name);
+			if (\is_array($value)) {
+				$getter = 'get' . ucfirst($name);
+				$currentValues = \is_array($defaultPolicy->$getter()) ? $defaultPolicy->$getter() : [];
+				$defaultPolicy->$setter(array_values(array_unique(array_merge($currentValues, $value))));
```
Comment (ChristophWurst) [inline lib/private/Security/FeaturePolicy/FeaturePolicyManager.php]:
```suggestion
				$defaultPolicy->$setter(\array_values(\array_unique(\array_merge($currentValues, $value))));
```

or import them ;)

File: lib/public/AppFramework/Http/EmptyFeaturePolicy.php [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+ *
+ * As alternative with sane exemptions look at FeaturePolicy
+ *
+ * @see \OCP\AppFramework\Http\FeaturePolicy
+ * @package OCP\AppFramework\Http
+ * @since 17.0.0
+ */
+class EmptyFeaturePolicy {
+
+	/** @var array of allowed domains to autoplay media */
```
Comment (ChristophWurst) [inline lib/public/AppFramework/Http/EmptyFeaturePolicy.php]:
```suggestion
	/** @var string[] of allowed domains to autoplay media */
```

File: lib/public/AppFramework/Http/EmptyFeaturePolicy.php [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+	/** @var array of allowed domains that can use fullscreen */
+	protected $fullscreenDomains = null;
+
+	/** @var array of allowed domains that can use the geolocation of the device */
+	protected $geolocationDomains = null;
+
+	/** @var array of allowed domains that can use the microphone */
+	protected $microphoneDomains = null;
+
+	/** @var array of allowed domains that can use the payment API */
```
Comment (ChristophWurst) [inline lib/public/AppFramework/Http/EmptyFeaturePolicy.php]: same for all vars until here

**Review by ChristophWurst — COMMENTED** [review]
looks good otherwise

File: lib/private/AppFramework/DependencyInjection/DIContainer.php (hunk 7 above)
Comment (rullzer) [reply lib/private/AppFramework/DependencyInjection/DIContainer.php]: of course you are right.

**Comment by rullzer** [comment]
All done @ChristophWurst :wink:

**Review by ChristophWurst — APPROVED** [review]
looks good!

File: lib/private/AppFramework/Middleware/Security/FeaturePolicyMiddleware.php [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+	 * Performs the default FeaturePolicy modifications that may be injected by other
+	 * applications
+	 *
+	 * @param Controller $controller
+	 * @param string $methodName
+	 * @param Response $response
+	 * @return Response
+	 */
+	public function afterController($controller, $methodName, Response $response): Response {
+		$policy = !is_null($response->getFeaturePolicy()) ? $response->getFeaturePolicy() : new FeaturePolicy();
```
Comment (kesselb) [inline lib/private/AppFramework/Middleware/Security/FeaturePolicyMiddleware.php]:
```suggestion
		$policy = $response->getFeaturePolicy() ?? new FeaturePolicy();
```

Could that work?

Comment (rullzer) [reply lib/private/AppFramework/Middleware/Security/FeaturePolicyMiddleware.php]: Probably. But lets leave it for now.

**Review by georgehrke — APPROVED** [review]
👍 Tested and works

**Review by danxuliu — APPROVED** [review]
Tested (with Talk) and works :+1:

**Comment by rullzer** [comment]
Pushed tests. Will merge once CI is happy

**Pull request merged**